packages/tic-tac-toe/app/controllers/game.ts

```diff
@@ -28,19 +49,119 @@ export default class GameController extends Controller {
+        if (this.currentGame.getPlayer() === Player.B) {
+          if (ChannelState.isRunning(channelState)) {
+            if (ChannelState.inXPlaying(channelState)) {
+              this.player = 'o';
+              this.Xs = channelState.appData.Xs;
+              this.Os = channelState.appData.Os;
+              console.log('Moving from xPlaying -> oPlaying');
+            }
+            if (ChannelState.inOPlaying(channelState)) {
+              this.player = 'x';
+              this.Xs = channelState.appData.Xs;
+              this.Os = channelState.appData.Os;
+              console.log('Moving from oPlaying -> xPlaying');
+            }
+          }
+        } else {
+          if (ChannelState.isRunning(channelState)) {
+            if (ChannelState.inXPlaying(channelState)) {
+              this.player = 'o';
+              this.Xs = channelState.appData.Xs;
+              this.Os = channelState.appData.Os;
+              console.log('Moving from xPlaying -> oPlaying');
+            }
+            if (ChannelState.inOPlaying(channelState)) {
+              this.player = 'x';
+              this.Xs = channelState.appData.Xs;
+              this.Os = channelState.appData.Os;
+              console.log('Moving from oPlaying -> xPlaying');
+            }
+          }
+        }
+      }
+    );
+  }
+
   @action
   protected setTile(index: number): void {
-    if ((1 << index) & this.Xs || (1 << index) & this.Os) {
+    if ((1 << index) & this.Xs || (1 << index) & this.Os || !this.isItMyTurn) {
       return;
     }
 
     if (this.player == 'x') {
       this.Xs |= 1 << index;
-      this.player = 'o';
+      const currentChannelState = this.currentGame.getChannelState();
+      const xPlayingState: AppData = {
+        type: 'xPlaying',
+        stake: this.currentGame.getGame().stake,
+        Xs: this.Xs,
+        Os: this.Os
+      };
+
+      this.tttChannelClient.updateChannel(
+        currentChannelState.channelId,
+        currentChannelState.aAddress,
+        currentChannelState.bAddress,
+        currentChannelState.aBal,
+        currentChannelState.bBal,
+        xPlayingState,
+        currentChannelState.aOutcomeAddress,
+        currentChannelState.bOutcomeAddress
+      );
+      console.log('Setting X on board');
     } else {
       this.Os |= 1 << index;
-      this.player = 'x';
+      const currentChannelState = this.currentGame.getChannelState();
+      const oPlayingState: AppData = {
+        type: 'oPlaying',
+        stake: this.currentGame.getGame().stake,
+        Xs: this.Xs,
+        Os: this.Os
+      };
+
+      this.tttChannelClient.updateChannel(
+        currentChannelState.channelId,
+        currentChannelState.aAddress,
+        currentChannelState.bAddress,
+        currentChannelState.aBal,
+        currentChannelState.bBal,
+        oPlayingState,
+        currentChannelState.aOutcomeAddress,
+        currentChannelState.bOutcomeAddress
+      );
+      console.log('Setting O on board');
```

Comment:
this looks really, really similar to [this](https://github.com/statechannels/monorepo/pull/1725/files#diff-1fb4bd5d34116c45523adad2892b02ecR107-R125). Couldn't it be refactored to a separate function that sets the state and updates the channel?

Reply:
Done thanks!